datafusion/ffi/src/execution/task_ctx.rs

```diff
@@ -0,0 +1,281 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{ffi::c_void, sync::Arc};
+
+use abi_stable::{
+    pmr::ROption,
+    std_types::{RHashMap, RString},
+    StableAbi,
+};
+use datafusion_execution::{config::SessionConfig, runtime_env::RuntimeEnv, TaskContext};
+use datafusion_expr::{
+    AggregateUDF, AggregateUDFImpl, ScalarUDF, ScalarUDFImpl, WindowUDF, WindowUDFImpl,
+};
+
+use crate::{
+    session_config::FFI_SessionConfig, udaf::FFI_AggregateUDF, udf::FFI_ScalarUDF,
+    udwf::FFI_WindowUDF,
+};
+
+/// A stable struct for sharing [`TaskContext`] across FFI boundaries.
+#[repr(C)]
+#[derive(Debug, StableAbi)]
+#[allow(non_camel_case_types)]
+pub struct FFI_TaskContext {
+    pub session_id: unsafe extern "C" fn(&Self) -> RString,
+
+    pub task_id: unsafe extern "C" fn(&Self) -> ROption<RString>,
+
+    pub session_config: unsafe extern "C" fn(&Self) -> FFI_SessionConfig,
+
+    pub scalar_functions: unsafe extern "C" fn(&Self) -> RHashMap<RString, FFI_ScalarUDF>,
+
+    pub aggregate_functions:
+        unsafe extern "C" fn(&Self) -> RHashMap<RString, FFI_AggregateUDF>,
+
+    pub window_functions: unsafe extern "C" fn(&Self) -> RHashMap<RString, FFI_WindowUDF>,
+
+    /// Release the memory of the private data when it is no longer being used.
+    pub release: unsafe extern "C" fn(arg: &mut Self),
+
+    /// Internal data. This is only to be accessed by the provider of the plan.
+    /// The foreign library should never attempt to access this data.
+    pub private_data: *mut c_void,
```

Comment:
I am not sure, but could this lead to a double free? for example, if FFI users clone the FFI_TaskContext, we might end up holding multiple references to `private_data` .

Reply:
I was thinking on explicitly forbidding cloning / copying, somthing like 
```
impl !Clone for FFI_TaskContext {} 
impl !Copy for FFI_TaskContext {} 
```
I was thinking that worst case scenario , FFI users could do a memcopy of the struct , so that it would be somehow  "safer" to have something like 
```
pub struct FFI_TaskContext {
    // An opaque pointer to an Arc<TaskContextPrivateData>
    arc_ptr: *const Arc<TaskContextPrivateData>,
}
```
and then returning a raw pointer to inner data , something along these lines
```
    pub fn as_opaque(&self) -> *const c_void {
        self.arc_ptr as *const c_void
    }
```